fix(material/tree): add levelAccessor, childrenAccessor, TreeKeyManag…   …er; a11y and docs improvements (#27626)
angular/components: src/cdk/a11y/a11y.md

this.keyManager = new FocusKeyManager(...).withWrap();
```

#### Types of list key managers

> There are two varieties of `ListKeyManager`, `FocusKeyManager` and `ActiveDescendantKeyManager`.

Reviewer: Should this sentence also be updated?

Each item must also have an ID bound to the listbox's or menu's `aria-activedescendant`.

### TreeKeyManager

> `TreeKeyManager` manages the active option in a tree view. This is intended to be used with
> components that correspond to a `role="tree"` pattern.

Reviewer: ```suggestion
`TreeKeyManager` manages the active option in a tree view. USe this key manager for
components that implement a `role="tree"` pattern.
```

#### Basic usage

> Any component that uses a `TreeKeyManager` will generally do three things:

Reviewer: ```suggestion
Any component that uses a `TreeKeyManager` should do three things:
```
nit: technical docs should already be present tense. Also replaced "generally do" with the stronger  "should" (which I think is accurate here)

* Forward keyboard events from the managed component to the `TreeKeyManager` via `onKeydown`.

Each tree item should implement the `TreeKeyManagerItem` interface:
```ts
interface TreeKeyManagerItem {
  /** Whether the item is disabled. */
  isDisabled?: (() => boolean) | boolean;

  /** The user-facing label for this item. */
  getLabel?(): string;

  /** Perform the main action (i.e. selection) for this item. */
  activate(): void;

  /** Retrieves the parent for this item. This is `null` if there is no parent. */
  getParent(): TreeKeyManagerItem | null;

  /** Retrieves the children for this item. */
  getChildren(): TreeKeyManagerItem[] | Observable<TreeKeyManagerItem[]>;

  /** Determines if the item is currently expanded. */
  isExpanded: (() => boolean) | boolean;

  /** Collapses the item, hiding its children. */
  collapse(): void;

  /** Expands the item, showing its children. */
  expand(): void;

  /**
   * Focuses the item. This should provide some indication to the user that this item is focused.
   */
  focus(): void;
}
```

#### Focus management

The `TreeKeyManager` will handle focusing the appropriate item on keyboard interactions. However,

> the component should call `onInitialFocus` when the component is focused for the first time (i.e.

Author: MatTree implements the [WAI ARIA Tree View Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/). We intend to follow the keyboard behaviors in that document:
 - "If none of the nodes are selected before the tree receives focus, focus is set on the first node."
 - "If a node is selected before the tree receives focus, focus is set on the selected node."

`TreeKeyManager#onInitialFocus` is an implementation detail of achieve this behavior. When MatTree is focused for the very first time, it calls onInitialFocus. If needed, the key manager sets focus to the first node. `onInitialFocus` is an implementation detail that I think 90% of developers will not need to know about. All most use cases will need to know is that when the tree is focused for the first time, it picks the first node to focus.

If you are making your own key manager, then you need to be familiar with `TreeKeyManagerStrategy#onInitialFocus`. The contract is that MatTree calls this method the very first time it is focused. It is up to the key manager how to handle this. The default keymanager `TreeKeyManagerStrategy` implements the WAI ARIA Tree View Pattern (focus first node if no node has ever been focused before).

I don't think `onInitialFocus` needs to be mentioned in this part of the documentation.

The important part is that MatTree manages it's own focus. Inject a TreeKeyManager to change focus behavior.

> the component should call `onInitialFocus` when the component is focused for the first time (i.e.

Reviewer 2: re: tabIndex; currently we do require MatTree (or other tree implementations using the key manager) to manage tabIndex
